iceoryx_utils/include/iceoryx_utils/internal/cxx/forward_list.inl

```diff
@@ -0,0 +1,581 @@
+{
+    return iterator{this, BEFORE_BEGIN_USED_INDEX};
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::before_begin() const noexcept
+{
+    return cbefore_begin();
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::cbefore_begin() const noexcept
+{
+    return const_iterator{this, BEFORE_BEGIN_USED_INDEX};
+}
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::begin() noexcept
+{
+    iterator iter = before_begin();
+    return ++iter;
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::begin() const noexcept
+{
+    return cbegin();
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::cbegin() const noexcept
+{
+    const_iterator citer = cbefore_begin();
+    return ++citer;
+}
+
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::end() noexcept
+{
+    return iterator{this, INVALID_INDEX};
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::end() const noexcept
+{
+    return cend();
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::cend() const noexcept
+{
+    return const_iterator{this, INVALID_INDEX};
+}
+
+
+template <typename T, int64_t CAPACITY>
+inline bool forward_list<T, CAPACITY>::empty() const noexcept
+{
+    return (m_size == 0);
+}
+
+template <typename T, int64_t CAPACITY>
+inline bool forward_list<T, CAPACITY>::full() const noexcept
+{
+    return (m_size >= CAPACITY);
+}
+
+template <typename T, int64_t CAPACITY>
+inline typename forward_list<T, CAPACITY>::size_type forward_list<T, CAPACITY>::size() const noexcept
+{
+    return m_size;
+}
+
+template <typename T, int64_t CAPACITY>
+inline typename forward_list<T, CAPACITY>::size_type forward_list<T, CAPACITY>::capacity() const noexcept
+{
+    return CAPACITY;
+}
+
+template <typename T, int64_t CAPACITY>
+inline typename forward_list<T, CAPACITY>::size_type forward_list<T, CAPACITY>::max_size() const noexcept
+{
+    return capacity();
+}
+
+
+template <typename T, int64_t CAPACITY>
+template <typename... ConstructorArgs>
+bool forward_list<T, CAPACITY>::emplace_front(ConstructorArgs&&... args) noexcept
+{
+    return (cend() != emplace_after(before_begin(), std::forward<ConstructorArgs>(args)...));
+}
+
+template <typename T, int64_t CAPACITY>
+template <typename... ConstructorArgs>
+typename forward_list<T, CAPACITY>::iterator
+forward_list<T, CAPACITY>::emplace_after(const_iterator afterToBeEmplacedIter, ConstructorArgs&&... args) noexcept
+{
+    if (this != afterToBeEmplacedIter.m_cList)
```

Comment:
I am torn about stuff like this. On one hand, it makes certain misuses very visible, via terminate (disregard for a moment whether terminate is justified or an error message would suffice, suitable error levels are another matter). On the other hand, assuming it is used right, we always have a runtime check here. I prefer asserts for this reason, errors are visible in debug mode but no checks are performed in release mode.